### PR TITLE
[release/7.0] Remove ConcurrentDictionary from the shared JsonSerializerOptions cache implementation.

### LINKED ISSUE #80258: Memory leak in blazor server application .net 7
### Description

as planned .net 7 upgrade on our existing blazor projects, we noticed memory usage on performance test increased drastically, with dockerized container we can drive the application to OOM exception. 
reproduced and profiled locally, it shows the top locking objects are ConcurrentDictionary+Tables<JsonSerializerOptions, WeakReference<JsonSerializerOptions+CachingContext>>
https://github.com/dotnet/runtime/issues/76548 - this issue was already reported in November 2022. 

this is blocking us from upgrading an important application in production to .net 7. I have attached the sample project to demonstrate the OOM exception, with GCHeapHardLimit set to 1GB, 10+ webpage reload of blazor application will get OOM exception. 

### Reproduction Steps

Launch blazor server application attached, open http://localhost:5000 to load the sample application, refresh the page and observe the memory increasing with Generation2 stack and not be able to free after GC.

using the HeapHardLimit value in the sample (1GB), observe OOM exception in about 10 reloads

### Expected behavior

Memory properly GCed

### Actual behavior

OOM exception occurs

### Regression?

this is new in .net 7 with the C

### Known Workarounds

no workarounds

### Configuration

.net 7
it happens in all OS
[telerikgridhorscrollingapp.zip](https://github.com/dotnet/runtime/files/10355267/telerikgridhorscrollingapp.zip)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fixes #80258

Backport of #76607, #76782

# Customer Impact

.NET 7 saw the release of a [global `JsonSerializerOptions` metadata cache](https://github.com/dotnet/runtime/pull/64646), intended to improve performance for users that create transient `JsonSerializerOptions` instances in their serialization operations. The implementation uses a `ConcurrentDictionary` keyed on `JsonSerializerOptions` configuration points, pointing to weak references of metadata caches. The cache uses a manual eviction strategy that removes dangling dictionary entries, triggered on every deserialization operation.

What this approach failed to foresee is that JsonSerializerOptions configuration points can be fairly large in their own right. In particular, Blazor uses transient `JsonSerializerOptions` instances as a vehicle for performing dependency injection in their custom converters. As a result, the custom converters that Blazor uses capture a _substantial amount of memory that gets rooted by the global cache_. Even though these keys will eventually get evicted, when this happens is predicated on the frequency of subsequent serialization operations, so in certain cases this can trigger bounded memory leaks.

The fix involves removing the ConcurrentDictionary and replacing it with a fixed-size array of weak references that are looked up using linear traversal. This removes the need for manual eviction and simplifies the cache implementation. Even though this was fixed in .NET 8 back in October, we've received multiple Blazor customer reports about this issue, convincing us it should be backported to .NET 7.

# Testing

Manually validated that memory leaks no longer occur in reproducing Blazor apps. Ran benchmarks confirming that global cache performance hasn't regressed.

# Risk

Moderate. It is an involved rewrite of a cache implementation, but it has been present in the .NET 8 branch for the past 3 months.

### CODE REVIEW

**Comment by carlossanlop** [comment]
For clarity, this PR does not need OOB changes added because it will go in the same release as this other PR for the same assembly, which already includes the required OOB changes: https://github.com/dotnet/runtime/pull/80513/files

**Comment by carlossanlop** [comment]
@eiriktsarpalis can you please solve the merge conflict? It showed up after merging the other Json PR.

**Comment by eiriktsarpalis** [comment]
Changes rebased.

**Comment by carlossanlop** [comment]
Approved by Tactics via email (7.0.3).
CI failure is a networking issue that I've seen in other backports since yesterday: https://github.com/dotnet/runtime/issues/80578
The required OOB changes are already in (see https://github.com/dotnet/runtime/pull/80576#issuecomment-1381033998).
Need a code review sign-off from an area owner. @layomia / @krwq?

**Review by krwq — APPROVED** [review]

**Pull request merged**
